Part2 (#53)
NCAR/python-tutorial: source/tutorials/yourfirst.rst

      # Data types for each column (only if non-string)
      types = {'tempout':float, 'humout':float, 'dewpt':float}

   ..

   And finally, make changes to the function calls:

   .. code-block:: python
      :lineno-start: 14

      # Compute the dew point temperature
      dewpointtemp = [compute_dewpoint(t, h) for t, h in zip(data['tempout'], data['humout'])]

      # Output comparison of data
      print_comparison('DEW PT', data['date'], data['time'], data['dewpt'], dewpointtemp)

   ..

7. Git add and commit:

   .. code-block:: bash

      $ git add dewpointtempcomp.py
      $ git commit -m "Computed dew point temperature"

   ..

8. Let's learn more about the math module!

   Hopefully since you already imported code from your :code:`readdata`\, :code:`printing`, and :code:computation` modules, importing from a built-in package :code:`math` seemed a little less intimidating.

   So far you have only used the :code:`math.log` function, but let's test out some other common methods within :code:`math`\.

   Perhaps you want to change the base of your logarithm. To do this you could type :code:`math.log(x, base)`. Here :code:`base` is

>    a keyword argument (just like :code:`filename` or :code:`types` in our :code:`read_data()` function) which means that :code:`base` does not need to be specified- 
>    when it is not the logarithm is assumed to be natural (base *e*). When both arguments are entered, the function returns the logarithm of :code:`x` to the given :code:`base`\, 

Reviewer: ```suggestion
   a keyword argument (just like :code:`filename` or :code:`types` in our :code:`read_data()` function) which
   means that :code:`base` does not need to be specified.  When it is not specified, the logarithm is assumed to
   be natural (base *e*). When both arguments are entered, the function returns the logarithm of :code:`x` to
   the given :code:`base`\, 
```

   calculated by :code:`log(x)/log(base)`\. Let's test this out:

   .. code-block:: python

      import math as m

>       x = math.e

Reviewer: ```suggestion
      x = m.e
```

      y_natural = m.log(x)
      y_base10 = m.log(x, 10)

      print(x, y_natural, y_base10)

   ..

>    Something new that we have done here is use the :code:`import ... as` statements. This essentially allows us to 

Reviewer: ```suggestion
   Something new that we have done here is use the \":code:`import ... as ...`\" statement. This essentially allows us to 
```

   shorten the name of the module for convenience if it is very long or if we are going to be calling it a lot.

>    :code:`math.e` calls up Euler's number (*e*), the base of the natural logarithm. *e* is an irrational number with infinite decimal places, often approximated as 2.718. 

Reviewer: ```suggestion
   The symbol :code:`math.e` represents Euler's number (*e*), the base of the natural logarithm. Euler's number (*e*) is an irrational number with infinite decimal places, often approximated as 2.718. 
```

   How much more accurate is :code:`math.e` than this approximation?
   

>    The above function :code:`math.log(x, base)` is very useful for computing logarithms in any base - but for some common bases there are separate logarithmic functions. 

Reviewer: ```suggestion
   The function :code:`math.log(x, base)` is very useful for computing logarithms in any base - but
   for some common bases there are separate logarithmic functions. 
```

   Try using :code:`log10(x)`\:

> 
> 

Reviewer: ```suggestion

```

   .. code-block:: python

      import math as m

      x = m.e

      y_natural = m.log(x)
      y_base10 = m.log(x, 10)
      y_log10 = m.log10(x)

      print(x, y_natural, y_base10, y_log10)

   ..

>    Do the two values differ? :code:`math.log10(x)` is considered to be more accurate than :code:`math.log(x, 10)`\.

Reviewer: ```suggestion
   Do the two values differ? The :code:`math.log10(x)` function is considered to be more accurate than :code:`math.log(x, 10)`\.
```

   Similarly :code:`math.log2(x)` is more accurate than :code:`math.log(x, 2)`\.

9. Let's cover some :code:`math` trigonometry examples!

>    Pi is an irrational number (like *e*) that is approximately :math:`\frac{22}{7}` or 3.14159. 

Reviewer: ```suggestion
   The math symbol :math:`\pi` is an irrational number (like *e*) that is approximately :math:`\frac{22}{7}` or 3.14159. 
```

   We can access the most accurate :code:`float` version of this number (depending on your C compiler), with :python:`math.pi`\.

   Say we wanted to convert a number from 60 degrees to radians. We have two options:

   .. code-block:: python

      import math as m

      deg = 60

      rads = deg * m.pi / 180

>       rads_fromfunc = radians(deg)

Reviewer: ```suggestion
      rads_fromfunc = m.radians(deg)
```

>       print(deg, m.pi, ras, rads_fromfunc)

Reviewer: ```suggestion
      print(deg, m.pi, rads, rads_fromfunc)
```

             time: List of strings representing time of day for each data element
             original_data: List of original data (floats)
             computed_data: List of computed data (floats)
          """

          print(f'                ORIGINAL  COMPUTED')
          print(f' DATE    TIME  {name.upper():>9} {name.upper():>9} DIFFERENCE')
          print(f'------- ------ --------- --------- ----------')
          for date, time, orig, comp in zip(date, time, original_data, computed_data):
             print(f'{date} {time:>6} {orig:9.6f} {comp:9.6f} {orig-comp:10.6f}')
    ..

    The only new functionality shown here is 
    :code:`string.upper()` (or, specifically, :code:`name.upper()`\), which capitalizes all lower case 
    letters in a string.

20. Edit the two scripts to use this new module (similar methods to step #12-14), and test your results.

    Try to do this on your own first, but if you are getting error messages the solution looks like:

    1)  Add the \":code:`from printing import print_comparison`\" line to the top of each script.

    2) Replace the printing output section at the bottom of each script with:

>     .. code-block:: python
>        :lineno-start: 29
> 
>        # Output comparison of data
>        print_comparison('WINDCHILL', data['date'], data['time'], data['windchill'], windchill)
> 
>     ..
> 
>     or
> 
>     .. code-block:: python
>        :lineno-start: 37
> 
>        # Output comparison of data
>        print_comparison('HEAT INDX', data['date'], data['time'], data['heatindex'], heatindex)
> 
>     ..

Reviewer: ```suggestion
       .. code-block:: python
          :lineno-start: 29

          # Output comparison of data
          print_comparison('WINDCHILL', data['date'], data['time'], data['windchill'], windchill)

       ..

       or

       .. code-block:: python
          :lineno-start: 37

          # Output comparison of data
          print_comparison('HEAT INDX', data['date'], data['time'], data['heatindex'], heatindex)

       ..
```

Reviewer: Missed this formatting

    by adding your import statements \":code:`from computation import compute_windchill`\"
    OR \":code:`from computation import compute_heatindex`\" and
    removing the redundant function definitions.

    Your two scripts should look as follows:

    For :code:`windchillcomp.py`\:

    .. code-block:: python
       :lineno-start: 1

       from readdata import read_data
       from printing import print_comparison
       from computation import compute_windchill

       # Column names and column indices to read
       columns = {'date':0, 'time':1, 'tempout':2, 'windspeed':7, 'windchill':12}

       # Data types for each column (only if non-string)
       types = {'tempout': float, 'windspeed':float, 'windchill':float}

       # Read data from file
       data = read_data(columns, types=types)

       # Compute the wind chill factor
       windchill = []
       for temp, windspeed in zip(data['tempout'], data['windspeed']):
          windchill.append(compute_windchill(temp, windspeed))

       # Output comparison of data
       print_comparison('WINDCHILL', data['date'], data['time'], data['windchill'], windchill)

    ..

> And for :code:`heatindexcomp.py`\:

Reviewer: Missed this formatting mistake, too...

```suggestion
    And for :code:`heatindexcomp.py`\:
```

   means that :code:`base` does not need to be specified.  When it is not specified, the logarithm is assumed to
   be natural (base *e*). When both arguments are entered, the function returns the logarithm of :code:`x` to
   the given :code:`base`\, 
   calculated by :code:`log(x)/log(base)`\. Let's test this out:

   .. code-block:: python

      import math as m

      x = m.e

      y_natural = m.log(x)
      y_base10 = m.log(x, 10)

      print(x, y_natural, y_base10)

   ..

   Something new that we have done here is use the \":code:`import ... as ...`\" statement. This essentially allows us to 
   shorten the name of the module for convenience if it is very long or if we are going to be calling it a lot.

   The symbol :code:`math.e` represents Euler's number (*e*), the base of the natural logarithm. Euler's number (*e*) is an irrational number with infinite decimal places, often approximated as 2.718. 
   How much more accurate is :code:`math.e` than this approximation?
   
   The function :code:`math.log(x, base)` is very useful for computing logarithms in any base - but
   for some common bases there are separate logarithmic functions. 
   Try using :code:`log10(x)`\:

>    .. code-block:: python

Reviewer: Needs a blank line before the Python code block.

```suggestion

   .. code-block:: python
```
